lang/src/org/partiql/lang/prettyprint/ASTPrettyPrinter.kt

```diff
@@ -0,0 +1,738 @@
+            is PartiqlAst.Statement.Exec -> toRecursionTree(ast)
+        }
+
+        return recursionTree.convertToString()
+    }
+
+    // ********
+    // * EXEC *
+    // ********
+    private fun toRecursionTree(node: PartiqlAst.Statement.Exec): RecursionTree =
+        RecursionTree(
+            astType = "Exec",
+            children = listOf(
+                toRecursionTree(node.procedureName, "procedureName")
+            ) + node.args.mapIndexed { index, expr -> toRecursionTree(expr, "arg${index + 1}") }
+        )
+
+    // *******
+    // * DDL *
+    // *******
+    private fun toRecursionTree(node: PartiqlAst.Statement.Ddl): RecursionTree =
+        RecursionTree(
+            astType = "Ddl",
+            children = listOf(
+                toRecursionTree(node.op, "op")
+            )
+        )
+
+    private fun toRecursionTree(node: PartiqlAst.DdlOp, attrOfParent: String? = null): RecursionTree =
+        when (node) {
+            is PartiqlAst.DdlOp.CreateIndex -> RecursionTree(
+                astType = "CreateIndex",
+                attrOfParent = attrOfParent,
+                children = listOf(
+                    toRecursionTree(node.indexName, "indexName")
+                ) + node.fields.mapIndexed { index, expr -> toRecursionTree(expr, "field${index + 1}") }
+            )
+            is PartiqlAst.DdlOp.CreateTable -> RecursionTree(
+                astType = "CreateTable",
+                attrOfParent = attrOfParent,
+                children = listOf(
+                    toRecursionTree(node.tableName, "tableName")
+                )
+            )
+            is PartiqlAst.DdlOp.DropIndex -> RecursionTree(
+                astType = "DropIndex",
+                attrOfParent = attrOfParent,
+                children = listOf(
+                    toRecursionTree(node.table, "table"),
+                    toRecursionTree(node.keys, "keys")
+                )
+            )
+            is PartiqlAst.DdlOp.DropTable -> RecursionTree(
+                astType = "DropTable",
+                attrOfParent = attrOfParent,
+                children = listOf(
+                    toRecursionTree(node.tableName, "tableName")
+                )
+            )
+        }
+
+    private fun toRecursionTree(node: PartiqlAst.Identifier, attrOfParent: String? = null): RecursionTree =
+        RecursionTree(
+            astType = "Identifier",
+            value = node.name.text + " " + node.case.toString(),
+            attrOfParent = attrOfParent
+        )
+
+    // *******
+    // * DML *
+    // *******
+    private fun toRecursionTree(node: PartiqlAst.Statement.Dml): RecursionTree =
+        RecursionTree(
+            astType = "Dml",
+            children = listOf(
+                toRecursionTree(node.operations, "operations")
+            ).let {
+                if (node.from == null) it else (it.plusElement(toRecursionTree(node.from!!, "from")))
```

Comment:
Curious, is using `!!` the best choice here and other similar situations?

Reply:
I think it is not necessary to add `!!` there, as Kotlin can smart cast it to non-null value. I will remove it.